clippy_lints/src/raw_strings.rs

```diff
@@ -0,0 +1,124 @@
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let r = "Hello, world!";
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_RAW_STRING,
+    restriction,
+    "suggests using a string literal when a raw string literal is unnecessary"
+}
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for raw string literals with an unnecessary amount of hashes around them.
+    ///
+    /// ### Why is this bad?
+    /// It's just unnecessary, and makes it look like there's more escaping needed than is actually
+    /// necessary.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let r = r###"Hello, "world"!"###;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let r = r#"Hello, "world"!"#;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_RAW_STRING_HASHES,
+    style,
+    "suggests reducing the number of hashes around a raw string literal"
+}
+impl_lint_pass!(RawStrings => [NEEDLESS_RAW_STRING, NEEDLESS_RAW_STRING_HASHES]);
+
+pub struct RawStrings {
+    pub needless_raw_string_hashes_allow_one: bool,
+}
+
+impl EarlyLintPass for RawStrings {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if !in_external_macro(cx.sess(), expr.span)
+            && let ExprKind::Lit(lit) = expr.kind
+            && let LitKind::StrRaw(max) | LitKind::ByteStrRaw(max) | LitKind::CStrRaw(max) = lit.kind
+        {
+            let prefix = match lit.kind {
+                LitKind::StrRaw(..) => "r",
+                LitKind::ByteStrRaw(..) => "br",
+                LitKind::CStrRaw(..) => "cr",
+                _ => unreachable!(),
+            };
+            if !snippet(cx, expr.span, prefix).trim().starts_with(prefix) {
+                return;
+            }
+
+            if !lit.symbol.as_str().contains(['\\', '"']) {
+                span_lint_and_sugg(
+                    cx,
+                    NEEDLESS_RAW_STRING,
+                    expr.span,
+                    "unnecessary raw string literal",
+                    "try",
+                    format!("{}\"{}\"", prefix.replace('r', ""), lit.symbol),
+                    Applicability::MachineApplicable,
+                );
+
+                return;
+            }
+
+            #[expect(clippy::cast_possible_truncation)]
+            let req = lit
+                .symbol
+                .as_str()
+                .as_bytes()
+                .split(|&b| b == b'"')
+                .skip(1)
+                .map_while(|bs| {
+                    let num = bs.iter().take_while(|&&b| b == b'#').count() as u8;
+                    if num.saturating_sub(1) == max {
+                        // Since `num - 1` is the max number of hashes it can have, we can
+                        // short-circuit with full confidence this is correct
+                        return None;
+                    }
+
+                    Some(num + 1)
+                })
+                .max()
+                .unwrap_or(0);
```

Comment:
```suggestion
            let req = if max == 0 {
                0
            } else {
                let mut following_quote = false;
                let mut cur_len = 0u8;
                let mut max_len = 0u8;
                for b in lit.symbol.as_str().as_bytes() {
                    match b {
                        b'"' => {
                            following_quote = true;
                            cur_len = 1;
                        },
                        b'#' => if following_quote {
                            cur_len += 1;
                        },
                        _ => if following_quote {
                            following_quote = false;
                            max_len = max_len.max(cur_len);
                            if max_len == max { break }
                        },
                    }
                }
                if following_quote {
                    // extra check in case string ends with hashes
                    max_len = max_len.max(cur_len)
                }
                max_len
            };
```
This is sort of what I had in mind. I had considered some kind of `fold`ing as well, but it would look even worse and short circuiting would be rough. Depends heavily on the benchmark of course, in the stuff I tested, I found a significant speedup compared to the other versions so far for strings with `max-1` hashes and little difference when included hashes were not `max-1`.

Reply:
This is indeed even faster, so thank you! But it's unreadable, so I think I'll at least try to reincorporate it into iterators.

Reply:
I've made it into an iterator, haven't benchmarked it yet but it has the same logic as yours. But at this point, it's both faster and more readable than the previous version so I think this is good enough, optimizing one lint to hell won't make or break clippy's performance :D